[TR3FIR-217] Add PKA controller with RSA functionalities

piotrkrygier-intent opened the pull request:
Created PKA driver for RSA operations. Operations include:
- Generating private exponent based on provided primes
- Encrypting message
- Decrypting message

Please note that due to the lack of dev board, development has been done on nucleo board, using mainline embassy. Some changes might be needed for this to work with our version of embassy.

Commit:
Implemented PKA module

Work in progres, there is a problem with exponent.

Fix modular exponent

Add finished PKA implementation with RSA operations

Fix constexpression calculations

Replace bytemuck with custom data conversion

alessandro-intent commented on embassy-stm32/src/pka/mod.rs:
```diff
@@ -0,0 +1,485 @@
+    /// Reorder data for the new word order
+    fn word_reorder<const N: usize>(input: &[u8; N]) -> [u8; N] {
+        // Ensure the input length is a multiple of 4 by padding with 0s if necessary
+        assert!(N % 4 == 0, "Input length must be a multiple of 4");
```
I think in general, whenever possible, we prefer to have an `Error` type and choose to return that without panicking, giving the option for the user to handle it in whichever way is appropriate.

alessandro-intent replied:
This applies to many parts of the code in these drivers, so I will not repeat the comments everywhere.

piotrkrygier-intent replied:
I understand. I will find parts of the code that might panic (like `unwrap`) and convert it to `Errror` return type.

piotrkrygier-intent replied:
It does not

alessandro-intent commented on embassy-stm32/src/pka/mod.rs:
```diff
@@ -0,0 +1,485 @@
+        for i in 0..(N / 4) {
+            let src_index = (N / 4 - 1 - i) * 4;
+            let dst_index = i * 4;
+            reordered[dst_index..dst_index + 4].copy_from_slice(&input[src_index..src_index + 4]);
```
If something goes wrong with the indexing,`copy_from_slice` will fail. Can we maybe gate the slice grabbing with [slice::get(_)](https://doc.rust-lang.org/std/primitive.slice.html#method.get). Then on the options returned we can convert to an `Error` type and avoid the panic.

alessandro-intent commented on embassy-stm32/src/pka/mod.rs:
```diff
@@ -0,0 +1,485 @@
+        let mut u32_data = [0; RSA_KEY_SIZE / 4];
+        for i in 0..data.len() / 4 {
+            let byte = T::regs().ram(offset + i).read();
+            info!("Reading from RAM: {:x} at offset: {:x}", byte, offset + i);
```
Sounds like this ought to be a `trace!`

alessandro-intent reviewed (commented):
I left a first pass, awesome work BTW.

grochowski-withintent reviewed (commented):
Nice bits. Pushing partial review so that the ongoing discussions don't get outdated.

grochowski-withintent reviewed (commented):
Sharing another portion of nitpicks. The basis looks very solid though.

piotrkrygier-intent commented on embassy-stm32/src/pka/mod.rs:
```diff
@@ -0,0 +1,485 @@
+{
+    /// Create a new RSA structure
+    pub fn new(pka: &'c mut Pka<'d, T, RSA_KEY_SIZE>) -> Self {
+        pka.enable();
```
We do not, I have checked. There is no reason at all to disable PKA, as it does absolutely nothing, other than disabling registers for reading and writing.

piotrkrygier-intent commented on embassy-stm32/src/lib.rs:
```diff
@@ -3,6 +3,7 @@
 #![feature(array_chunks)]
 #![feature(slice_as_chunks)]
 #![feature(pointer_is_aligned_to)]
+#![feature(generic_const_exprs)]
```
Without it, I'm unable to create an array of length [RSA_KEY_SIZE / 4]. The workaround would be to not create that array at all, but to somehow insert u32 value every 4 bytes of u8 result array, but that might be more error prone.

piotrkrygier-intent replied:
Did some dirty work, removed [RSA_KEY_SIZE / 4] and removed problematic feature

piotrkrygier-intent commented on embassy-stm32/src/pka/mod.rs:
```diff
@@ -0,0 +1,485 @@
+            });
+        }
+        // Write 0 to next word
+        T::regs().ram(offset + u32_data.len()).write(|w| {
+            *w = 0;
+            *w
+        });
+    }
```
Sure, makes much more sense

piotrkrygier-intent commented on embassy-stm32/src/pka/mod.rs:
```diff
@@ -0,0 +1,485 @@
+        let u32_data: &[u32] = cast_slice(data);
+
+        for (i, byte) in u32_data.iter().enumerate() {
+            T::regs().ram(offset + i).write(|w| {
+                *w = *byte;
+                *w
+            });
```
Basically yes, but I wanted to use that functionality to differentiate write_to_ram from write_to_ram_u32. I know it doesn't make much sense, so I will probably change it.

piotrkrygier-intent commented on embassy-stm32/src/pka/mod.rs:
```diff
@@ -0,0 +1,485 @@
+    /// Generate RSA key pair
+    pub async fn generate_key_pair<const N: usize>(&mut self, prime_1: [u8; N], prime_2: [u8; N]) {
+        // Generate public and private keys
+        if prime_2.len() > RSA_KEY_SIZE / 2 || prime_1.len() > RSA_KEY_SIZE / 2 {
```
I tried that, but no matter what values I provide I get the compile error.

piotrkrygier-intent commented on embassy-stm32/src/pka/mod.rs:
```diff
@@ -0,0 +1,485 @@
+    }
+
+    /// Encrypt data
+    pub async fn encrypt<const N: usize>(&mut self, data: &[u8; N]) -> Result<[u8; RSA_KEY_SIZE], ()> {
```
Removing `Result` from the return type.

Thanks for comment about `no_std`, didn't know that. I will replace returning an array with filling provided slice. This feels more like C anyway :)

piotrkrygier-intent commented on embassy-stm32/src/pka/mod.rs:
```diff
@@ -0,0 +1,485 @@
+    }
+
+    /// Decrypt data
+    pub async fn decrypt<const N: usize>(&mut self, data: &[u8; N]) -> Result<[u8; RSA_KEY_SIZE], ()> {
```
Unfortunately, I get compile error here as well, no matter what check I do (i.e. failed for N < 4)

piotrkrygier-intent commented on embassy-stm32/src/pka/mod.rs:
```diff
@@ -0,0 +1,485 @@
+        T::regs().cr().modify(|w| w.set_start(true));
+        // Wait for results
+        poll_fn(|ctx| {
+            if T::regs().sr().read().addrerrf() {
```
Commented some of the more mysterious regs calls

RafalUlko commented on embassy-stm32/src/pka/mod.rs:
```diff
@@ -0,0 +1,503 @@
+        }
+    }
+
+    /// Reorder data for the new word order
+    fn word_reorder<const N: usize>(input: &[u8; N], result: &mut [u8; N]) {
+        // Ensure the input length is a multiple of 4 by padding with 0s if necessary
+
+        for i in 0..(N / 4) {
+            let src_index = (N / 4 - 1 - i) * 4;
+            let dst_index = i * 4;
+            result[dst_index..dst_index + 4].copy_from_slice(&input[src_index..src_index + 4]);
+        }
+    }
```
Math is unnecessary when you have Rust :stuck_out_tongue: 
```rust
input.chunks(4)
    .rev()
    .enumerate()
    .for_each(|(i, c)| result[(i*4)..(i*4+4)].copy_from_slice(c));
```

RafalUlko commented on embassy-stm32/src/pka/mod.rs:
```diff
@@ -0,0 +1,503 @@
+                &mut reordered_data,
+            )
+            .await;
+        Self::word_reorder(&mut reordered_data, result);
```
Do first arg need to be mutable?

piotrkrygier-intent replied:
No, it doesn't. Good catch!

piotrkrygier-intent commented on embassy-stm32/src/pka/mod.rs:
```diff
@@ -0,0 +1,503 @@
+// List of RAM addresses for data. Since embassy implementation of RAM access already moves base address to
+// 0x0400 and multiplies offset by 4 for 32-bit access, we have to take address from reference manual, subtract 0x0400
+// and divide it by 4 to get the correct offset.
+const OPERAND_LENGTH_ADDR: usize = 0x0001; // 0x0404
+const MODULUS_ADDR: usize = 0x0257; // 0x0D5C
+const MONT_PARAM_R2_MOD_N: usize = 0x0065; // 0x0594
+const OPERAND_A_ADDR: usize = 0x012D; // 0x08B4
+const OPERAND_B_ADDR: usize = 0x0191; // 0x0A44
+const ARITHMETIC_RESULT_ADDR: usize = 0x01F4; // 0x0BD0
+const EXPONENT_LENGTH_ADDR: usize = 0x0000; // 0x0400
+const EXPONENT_ADDR: usize = ARITHMETIC_RESULT_ADDR; // 0x0BD0
+const EXPONENTIATION_RESULT_ADDR: usize = 0x00C9; // 0x0724
```
I could, but I would rather include more consts here to make it more readable. Some addresses in documentation share name, but not the address, i.e.: "Operand A" in Modular addition operation has address `0x8B4`, but "Operand A" in Modular exponentiation is located in `0xA44`, which is the address of "Operand B" in Arithmetic Subtraction or "Modulus value n" in Modular Inversion.

RafalUlko replied:
I proposed an Enum because I thought they are more readable :laughing:  I think you can have same values for different names when you use Enum with `repr(u32)`, or you can have separate address Enum for particular usage. 

Treat this message as a suggestion, I'm not insisting that it should be done this way. I assume you know better what you're doing here, so I am marking this thread as resolved.

piotrkrygier-intent replied:
I am good with embedded systems, but I still have long way to go wit "Rust way" :). I.e., I didn't know you could have multiple enums with the same value. That's convenient!

RafalUlko replied:
I verified that, you can not have duplicated values in same enum, but mby you can define each address group as separate enum.

RafalUlko commented on embassy-stm32/src/pka/mod.rs:
```diff
@@ -16,67 +16,297 @@ static PKA_WAKER: AtomicWaker = AtomicWaker::new();
 // List of RAM addresses for data. Since embassy implementation of RAM access already moves base address to
 // 0x0400 and multiplies offset by 4 for 32-bit access, we have to take address from reference manual, subtract 0x0400
 // and divide it by 4 to get the correct offset.
+enum MontParamComputation {
+    ModulusLength,
+    Modulus,
+    Result,
+}
+
+impl MontParamComputation {
+    const MODULUS_LENGTH_ADDR: usize = (0x0404 - 0x0400) / 4;
+    const MODULUS_ADDR: usize = (0x0D5C - 0x0400) / 4;
+    const RESULT_ADDR: usize = (0x0594 - 0x0400) / 4;
+
+    fn value(&self) -> usize {
+        match self {
+            MontParamComputation::ModulusLength => Self::MODULUS_LENGTH_ADDR,
+            MontParamComputation::Modulus => Self::MODULUS_ADDR,
+            MontParamComputation::Result => Self::RESULT_ADDR,
+        }
+    }
+}
+
```
You can assign value to enum directly https://doc.rust-lang.org/reference/items/enumerations.html#implicit-discriminants

RafalUlko replied:
It can be even an expression like `(0x0404 - 0x0400) /4`

piotrkrygier-intent replied:
I know, check the PKA mode enum. But if I implement this method, I can just call it whenever I need a usize value, instead of casting it each time.

RafalUlko replied:
I'm not insisting :p You can consider implementing the `From<MontParamComputation> for usize` trait. Then you can just call `into()` instead of `value()` and it seems more Rusty way for me.

RafalUlko commented on embassy-stm32/src/pka/mod.rs:
```diff
@@ -0,0 +1,830 @@
+        let operand_size = 8 * (a.len() as u32);
+
+        self.write_to_ram_u32(ArithmeticMultiplication::OperandLength.value(), &operand_size);
+        self.write_to_ram(ArithmeticMultiplication::OperandA.value(), &a);
```
You can replace `&a` with just `a` here. 

There are more issues like that. Please run `cargo clippy -- -D warnings` in `embassy-stm32`.

RafalUlko committed:
[TR3FIR-217] Refactor PKA and remove RSA struct

[TR3FIR-217] Remove RSA struct from PKA driver

[TR3FIR-217] Add padding for PKA RAM writes

[TR3FIR-217] Fix modulus padding in PKA mod exp method

[TR3FIR-217] Fix modular exp in PKA

[TR3FIR-217] Fix Mongtomery param in PKA

mattcaron-token reviewed (commented):
This mostly looks fine. I just have a couple questions I'd like to talk through before I approve.

RafalUlko commented on embassy-stm32/src/pka/mod.rs:
```diff
@@ -0,0 +1,497 @@
+
+// List of RAM addresses for data. Since embassy implementation of RAM access already moves base address to
+// 0x0400 and multiplies offset by 4 for 32-bit access, we have to take address from reference manual, subtract 0x0400
+// and divide it by 4 to get the correct offset.
```
It would require changing `pka_v1c.rs` in `stm32-metapac` because there all that math is reveses.

mattcaron-token replied:
Yeah, let's not do that. ;-)

RafalUlko commented on embassy-stm32/src/pka/mod.rs:
```diff
@@ -0,0 +1,497 @@
+    }
+    /// Encrypt data
+    pub async fn encrypt(&mut self, data: &[u32]) -> Result<[u32; RSA_KEY_SIZE], ()> {
+        // Message to encrypt should be less than modulus
```
You are right about encryption len. Anyway RSA struct has been removed from PKA module. Sice IMO implementing RSA using PKA should be done in higher abstraction layer. In `embassy-stm32` should be done only access to peripheral.

RafalUlko commented:
Squashed all Piotr's and my commits into two separate ones.

mattcaron-token reviewed (approved):
I've reviewed most of this already, but I re-read through it and it looks fine.
